peerpod-ctrl: #375 follow up (#767)
confidential-containers/cloud-api-adaptor: peerpod-ctrl/README.md

# peerpod-ctrl
The PeerPod CR is used to track peer-pods cloud-provider resources, it specifies the InstanceID and its CloudProvider. PeerPod objects are owned by the matching (peer)Pod object.

> The PeerPod controller is watching PeerPod events and deleting hanging resources that were not deleted by the cloud-api-adaptor at Pod deletion time.

Reviewer: ```suggestion
The PeerPod CR is used to track the cloud provider resources for a Pod; it requires the cloud InstanceID and the  CloudProvider. PeerPod objects are owned by the matching (peer)Pod object.
The PeerPod controller is watching PeerPod events and deleting dangling resources that were not deleted by the cloud-api-adaptor at Pod deletion time.
```

## Description
### Creation time:

> With every successful peer-pod VM creation cloud-api-adaptor will create a PeePod CR (predefined by the operator) which contains the cloud provider and its VM instance id.

Reviewer: ```suggestion
With every successful VM creation for a Pod, cloud-api-adaptor will create a PeePod CR (predefined by the operator) which contains the cloud provider and its VM instance id
```

### Owner references:

> The PeerPod CR is owned by the original Pod object that represents the peer-pod, upon (peer)Pod deletion [background cascading deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#background-deletion) is getting into action, hence, the Pod will be deleted first and the GC will handle the owned PeerPod CR later.

Reviewer: ```suggestion
The PeerPod CR is owned by the original Pod object. Upon Pod deletion [background cascading deletion](https://kubernetes.io/docs/concepts/architecture/garbage-collection/#background-deletion) gets into action and hence the Pod will be deleted first, followed by GC handling the owned PeerPod CR.
```

### Deletion time:

> * Normal case: When remote hypervisor will get the stopVM request (upon (peer)Pod deletion) it will delete the podVM instance and if it succeeds it will remove the finalizer attached to the owned PeerPod object so it can then be cleaned by the GC

Reviewer: ```suggestion
Normal case: When remote hypervisor will get the stopVM request (upon Pod deletion) it will delete the pod VM instance and if it succeeds it will remove the finalizer attached to the owned PeerPod object so it can then be cleaned by the GC.
```

> * Failure case: If for any reason cloud-api-adaptor doesn’t honor the delete request or it fails to preform deletion, the finalizer is not removed, hence, when PeerPod controller gets a delete event for the owned PeerPod object by the GC but it still has the finalizer it will comprehend it needs to preform the deletion of podVM resource by itself, based on the PeerPod CR fields

Reviewer: ```suggestion
Failure case: If for any reason cloud-api-adaptor doesn’t honor the delete request or it fails to perform deletion, the finalizer is not removed. Hence, when PeerPod controller gets a delete event for the owned PeerPod object by the GC and it still has the finalizer, it will comprehend that it needs to perform the deletion of pod VM resource by itself, based on the PeerPod CR fields.
```
